File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        ListNode dummy = new ListNode();
        dummy.next = head;
        
        ListNode prev = dummy, curr = head, next;
        ListNode h = head;
        
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            if (curr == h) curr.next = null;
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
    
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    
    int getNodeCount(ListNode* head) {
        int res = 0;
        while (head != nullptr) {
            head = head->next;
            res++;
        }
        return res;
    }
    
    ListNode* reverseList(ListNode* head) {
        // Handle the edge case of an empty list
        if (!head) return nullptr;

        // Create a dummy node to handle the case when left = 1
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* prev = dummy;
        
        // Start reversing the sublist
        ListNode* current = prev->next;
        ListNode* next = nullptr;
        
        int n = getNodeCount(head);

        for (int i = 0; i < n - 1; i++) {
            next = current->next;
            current->next = next->next;
            next->next = prev->next;
            prev->next = next;
        }

        return dummy->next;
    }
    

};

